File: androidApp/src/main/java/org/multipaz/wallet/android/ui/document/CredentialInfoScreen.kt
package org.multipaz.wallet.android.ui.document

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.outlined.Delete
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.MediumTopAppBar
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.material3.rememberTopAppBarState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.input.nestedscroll.nestedScroll
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import org.multipaz.wallet.android.R
import org.multipaz.revocation.IdentifierList
import org.multipaz.revocation.RevocationCheckState
import org.multipaz.revocation.RevocationChecker
import org.multipaz.revocation.RevocationInfo
import org.multipaz.revocation.RevocationStatus
import org.multipaz.revocation.StatusList
import org.multipaz.revocation.getRevocationInfo
import org.multipaz.trustmanagement.TrustManagerInterface
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.launch
import kotlinx.io.bytestring.decodeToString
import kotlinx.serialization.json.jsonPrimitive
import org.multipaz.cbor.Cbor
import org.multipaz.cbor.Tagged
import org.multipaz.claim.JsonClaim
import org.multipaz.claim.MdocClaim
import org.multipaz.claim.organizeByNamespace
import org.multipaz.compose.datetime.formattedDateTime
import org.multipaz.compose.document.CredentialInfo
import org.multipaz.compose.document.DocumentModel
import org.multipaz.compose.items.FloatingItemCenteredText
import org.multipaz.compose.items.FloatingItemHeadingAndText
import org.multipaz.compose.items.FloatingItemList
import org.multipaz.cose.Cose
import org.multipaz.cose.CoseNumberLabel
import org.multipaz.credential.Credential
import org.multipaz.credential.SecureAreaBoundCredential
import org.multipaz.crypto.X509CertChain
import org.multipaz.mdoc.credential.MdocCredential
import org.multipaz.sdjwt.credential.SdJwtVcCredential
import org.multipaz.util.Logger
import org.multipaz.util.toHex
import kotlin.time.Duration
import org.multipaz.securearea.AndroidKeystoreKeyInfo
import org.multipaz.securearea.software.SoftwareKeyInfo
import org.multipaz.securearea.cloud.CloudKeyInfo
import org.multipaz.wallet.android.ui.Note

private const val TAG = "CredentialInfoScreen"

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CredentialInfoScreen(
    documentModel: DocumentModel,
    documentId: String,
    credentialId: String,
    revocationChecker: RevocationChecker,
    issuerTrustManager: TrustManagerInterface,
    onBackClicked: () -> Unit,
    onViewCertificateChain: (certChain: X509CertChain) -> Unit,
    showToast: (message: String) -> Unit
) {
    val coroutineScope = rememberCoroutineScope()
    var showDeleteConfirmationDialog by remember { mutableStateOf(false) }

    val documentInfos = documentModel.documentInfos.collectAsState().value
    val documentInfo = documentInfos.find { it.document.identifier == documentId }
    val credentialInfo = documentInfo?.credentialInfos?.find { it.credential.identifier == credentialId  }

    if (showDeleteConfirmationDialog) {
        AlertDialog(
            onDismissRequest = { showDeleteConfirmationDialog = false },
            title = {
                Text(text = stringResource(R.string.credential_info_delete_pending_title))
            },
            text = {
                Text(text = stringResource(R.string.credential_info_delete_pending_text))
            },
            dismissButton = {
                TextButton(
                    onClick = { showDeleteConfirmationDialog = false }
                ) {
                    Text(text = stringResource(R.string.credential_info_delete_pending_cancel))
                }
            },
            confirmButton = {
                TextButton(
                    onClick = {
                        showDeleteConfirmationDialog = false
                        coroutineScope.launch {
                            try {
                                documentInfo?.document?.deleteCredential(credentialId)
                                onBackClicked()
                            } catch (e: Throwable) {
                                if (e is CancellationException) throw e
                                Logger.e(TAG, "Failed to delete credential $credentialId", e)
                                showToast("Failed to delete credential: ${e.message}")
                            }
                        }
                    }
                ) {
                    Text(text = stringResource(R.string.credential_info_delete_pending_confirm))
                }
            }
        )
    }

    val scrollBehavior = TopAppBarDefaults.enterAlwaysScrollBehavior(rememberTopAppBarState())
    Scaffold(
        modifier = Modifier
            .nestedScroll(scrollBehavior.nestedScrollConnection)
            .fillMaxSize(),
        topBar = {
            MediumTopAppBar(
                title = {
                    Text(stringResource(R.string.credential_info_title))
                },
                navigationIcon = {
                    IconButton(onClick = onBackClicked) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                            contentDescription = null
                        )
                    }
                },
                actions = {
                    if (credentialInfo != null && !credentialInfo.credential.isCertified) {
                        IconButton(onClick = { showDeleteConfirmationDialog = true }) {
                            Icon(
                                imageVector = Icons.Outlined.Delete,
                                contentDescription = stringResource(R.string.credential_info_delete_pending_content_description)
                            )
                        }
                    }
                },
                scrollBehavior = scrollBehavior
            )
        }
    ) { padding ->
        Column(
            modifier = Modifier
                .padding(padding)
                .padding(16.dp)
                .fillMaxSize()
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            Note(
                markdownString = stringResource(R.string.credential_info_note)
            )
            if (credentialInfo != null) {
                FloatingItemList() {
                    CredentialInfoSection(
                        credentialInfo = credentialInfo,
                        revocationChecker = revocationChecker,
                        issuerTrustManager = issuerTrustManager,
                        onViewCertificateChain = onViewCertificateChain,
                        showToast = showToast
                    )
                }
                if (credentialInfo.credential.isCertified) {
                    CredentialClaimsSection(credentialInfo)
                }
            } else {
                FloatingItemList {
                    FloatingItemCenteredText(stringResource(R.string.credential_info_empty))
                }
            }
            Spacer(modifier = Modifier.height(20.dp))
        }
    }
}

@Composable
private fun CredentialInfoSection(
    credentialInfo: CredentialInfo,
    revocationChecker: RevocationChecker,
    issuerTrustManager: TrustManagerInterface,
    onViewCertificateChain: (certChain: X509CertChain) -> Unit,
    showToast: (message: String) -> Unit
) {
    val coroutineScope = rememberCoroutineScope()

    FloatingItemHeadingAndText(stringResource(R.string.credential_info_type), credentialInfo.credential.credentialType)
    FloatingItemHeadingAndText(stringResource(R.string.credential_info_identifier), credentialInfo.credential.identifier)
    FloatingItemHeadingAndText(stringResource(R.string.credential_info_domain), credentialInfo.credential.domain)
    FloatingItemHeadingAndText(stringResource(R.string.credential_info_certified), if (credentialInfo.credential.isCertified) stringResource(R.string.credential_info_yes) else stringResource(R.string.credential_info_no))
    if (credentialInfo.credential.isCertified) {
        FloatingItemHeadingAndText(
            stringResource(R.string.credential_info_valid_from),
            formattedDateTime(credentialInfo.credential.validFrom)
        )
        FloatingItemHeadingAndText(
            stringResource(R.string.credential_info_valid_until),
            formattedDateTime(credentialInfo.credential.validUntil)
        )
        FloatingItemHeadingAndText(
            "Issuer provided data",
            "${credentialInfo.credential.issuerProvidedData.size} bytes"
        )
        FloatingItemHeadingAndText("Usage Count", credentialInfo.credential.usageCount.toString())
        RevocationStatusSection(
            revocationChecker = revocationChecker,
            issuerTrustManager = issuerTrustManager,
            credential = credentialInfo.credential
        )
        when (credentialInfo.credential) {
            is MdocCredential -> {
                val issuerSigned = Cbor.decode(credentialInfo.credential.issuerProvidedData.toByteArray())
                val issuerAuth = issuerSigned["issuerAuth"].asCoseSign1
                val msoBytes = issuerAuth.payload!!
                FloatingItemHeadingAndText("MSO size", "${msoBytes.size} bytes")
                FloatingItemHeadingAndText(
                    "ISO mdoc DocType",
                    (credentialInfo.credential as MdocCredential).docType
                )
                FloatingItemHeadingAndText(
                    showChevron = true,
                    heading = "ISO mdoc DS Key Certificate",
                    text = buildAnnotatedString {
                        withStyle(style = SpanStyle(color = MaterialTheme.colorScheme.secondary)) {
                            append("Click for details")
                        }
                    },
                    modifier = Modifier.clickable {
                        coroutineScope.launch {
                            val certChain =
                                issuerAuth.unprotectedHeaders[
                                    CoseNumberLabel(Cose.COSE_LABEL_X5CHAIN)
                                ]!!.asX509CertChain
                            onViewCertificateChain(certChain)
                        }
                    }
                )
            }

            is SdJwtVcCredential -> {
                FloatingItemHeadingAndText(
                    "Verifiable Credential Type",
                    (credentialInfo.credential as SdJwtVcCredential).vct
                )
                // TODO: Show cert chain for key used to sign issuer-signed data. Involves
                //  getting this over the network as specified in section 5 "JWT VC Issuer Metadata"
                //  of https://datatracker.ietf.org/doc/draft-ietf-oauth-sd-jwt-vc/ ... how annoying
            }
        }
    }

    if (credentialInfo.credential is SecureAreaBoundCredential) {
        val keyInfo = credentialInfo.keyInfo
        FloatingItemHeadingAndText(
            "Secure Area",
            (credentialInfo.credential as SecureAreaBoundCredential).secureArea.displayName
        )
        FloatingItemHeadingAndText(
            "Secure Area Identifier",
            (credentialInfo.credential as SecureAreaBoundCredential).secureArea.identifier
        )
        if (keyInfo != null) {
            FloatingItemHeadingAndText(
                "Device Key Algorithm",
                keyInfo.algorithm.description
            )
            when (keyInfo) {
                is AndroidKeystoreKeyInfo -> {
                    FloatingItemHeadingAndText(
                        "Device Key in StrongBox",
                        if (keyInfo.isStrongBoxBacked) "Yes" else "No"
                    )
                    FloatingItemHeadingAndText(
                        "Device Key User Auth Required",
                        if (keyInfo.isUserAuthenticationRequired) "Yes" else "No"
                    )
                    if (keyInfo.isUserAuthenticationRequired) {
                        val timeoutText = if (keyInfo.userAuthenticationTimeout == Duration.ZERO) {
                            "0 (Authenticate for every use)"
                        } else {
                            "${keyInfo.userAuthenticationTimeout.inWholeSeconds} s"
                        }
                        FloatingItemHeadingAndText("Device Key User Auth Timeout", timeoutText)
                        val typesText = if (keyInfo.userAuthenticationTypes.isEmpty()) {
                            "None"
                        } else {
                            keyInfo.userAuthenticationTypes.joinToString(", ") { it.name }
                        }
                        FloatingItemHeadingAndText("Device Key User Auth Types", typesText)
                    }
                    if (keyInfo.attestKeyAlias != null) {
                        FloatingItemHeadingAndText("Device Key Attest Key Alias", keyInfo.attestKeyAlias!!)
                    }
                    if (keyInfo.validFrom != null) {
                        FloatingItemHeadingAndText("Device Key Valid From", formattedDateTime(keyInfo.validFrom!!))
                    }
                    if (keyInfo.validUntil != null) {
                        FloatingItemHeadingAndText("Device Key Valid Until", formattedDateTime(keyInfo.validUntil!!))
                    }
                }
                is SoftwareKeyInfo -> {
                    FloatingItemHeadingAndText(
                        "Device Key Passphrase Protected",
                        if (keyInfo.isPassphraseProtected) "Yes" else "No"
                    )
                    FloatingItemHeadingAndText(
                        "Device Key User Auth Required",
                        if (keyInfo.isUserAuthenticationRequired) "Yes" else "No"
                    )
                    if (keyInfo.isUserAuthenticationRequired) {
                        val typesText = if (keyInfo.userAuthenticationTypes.isEmpty()) {
                            "None"
                        } else {
                            keyInfo.userAuthenticationTypes.joinToString(", ") { it.name }
                        }
                        FloatingItemHeadingAndText("Device Key User Auth Types", typesText)
                    }
                }
                is CloudKeyInfo -> {
                    FloatingItemHeadingAndText(
                        "Device Key Passphrase Required",
                        if (keyInfo.isPassphraseRequired) "Yes" else "No"
                    )
                    FloatingItemHeadingAndText(
                        "Device Key User Auth Required",
                        if (keyInfo.isUserAuthenticationRequired) "Yes" else "No"
                    )
                    if (keyInfo.isUserAuthenticationRequired) {
                        val typesText = if (keyInfo.userAuthenticationTypes.isEmpty()) {
                            "None"
                        } else {
                            keyInfo.userAuthenticationTypes.joinToString(", ") { it.name }
                        }
                        FloatingItemHeadingAndText("Device Key User Auth Types", typesText)
                    }
                    if (keyInfo.validFrom != null) {
                        FloatingItemHeadingAndText("Device Key Valid From", formattedDateTime(keyInfo.validFrom!!))
                    }
                    if (keyInfo.validUntil != null) {
                        FloatingItemHeadingAndText("Device Key Valid Until", formattedDateTime(keyInfo.validUntil!!))
                    }
                }
            }
        }
        FloatingItemHeadingAndText("Device Key Invalidated",
            buildAnnotatedString {
                if (credentialInfo.keyInvalidated) {
                    withStyle(style = SpanStyle(
                        fontWeight = FontWeight.Bold,
                        color = MaterialTheme.colorScheme.error
                    )) {
                        append("YES")
                    }
                } else {
                    append("No")
                }
            })
        FloatingItemHeadingAndText(
            showChevron = true,
            heading = "Device Key Attestation",
            text = buildAnnotatedString {
                withStyle(style = SpanStyle(color = MaterialTheme.colorScheme.secondary)) {
                    append("Click for details")
                }
            },
            modifier = Modifier.clickable {
                coroutineScope.launch {
                    val attestation = (credentialInfo.credential as SecureAreaBoundCredential).getAttestation()
                    if (attestation.certChain != null) {
                        onViewCertificateChain(attestation.certChain!!)
                    } else {
                        showToast("No attestation for Device Key")
                    }
                }
            }
        )
    } else {
        FloatingItemHeadingAndText("Secure Area", "N/A")
    }
}

@Composable
private fun CredentialClaimsSection(credentialInfo: CredentialInfo) {
    when (credentialInfo.credential) {
        is MdocCredential -> {
            val mdocClaimsByNamespace = (credentialInfo.claims as List<MdocClaim>).organizeByNamespace()
            for ((namespace, claims) in mdocClaimsByNamespace) {
                FloatingItemList(title = "Namespace $namespace") {
                    claims.forEach { claim ->
                        FloatingItemHeadingAndText(
                            heading = claim.dataElementName,
                            text = claim.render()
                        )
                    }
                }
            }
        }
        else -> {
            FloatingItemList(title = "Claims") {
                credentialInfo.claims.forEach { claim ->
                    val claimName = if (claim is JsonClaim) {
                        claim.claimPath.map { it.jsonPrimitive.content }.joinToString(".")
                    } else {
                        claim.displayName
                    }
                    FloatingItemHeadingAndText(
                        heading = claimName,
                        text = claim.render()
                    )
                }
            }
        }
    }
}

@Composable
private fun RevocationStatusSection(
    revocationChecker: RevocationChecker,
    issuerTrustManager: TrustManagerInterface,
    credential: Credential
) {
    val coroutineScope = rememberCoroutineScope()
    val revocationData = remember { mutableStateOf<RevocationInfo?>(null) }

    LaunchedEffect(Unit) {
        coroutineScope.launch {
            revocationData.value = credential.getRevocationInfo(issuerTrustManager)
        }
    }
    val value = revocationData.value
    if (value != null) {
        RevocationCheckSection(revocationChecker, value)
    } else {
        FloatingItemHeadingAndText(
            heading = "Revocation info",
            text = "Not found"
        )
    }
}

@Composable
private fun RevocationCheckSection(
    revocationChecker: RevocationChecker,
    revocationData: RevocationInfo
) {
    val coroutineScope = rememberCoroutineScope()
    val statusText = remember { mutableStateOf("Click to check status") }
    val heading = when (revocationData.revocationStatus) {
        is RevocationStatus.StatusList -> "Status List Revocation"
        is RevocationStatus.IdentifierList -> "Identifier List Revocation"
        else -> "Revocation info"
    }
    val detailsText = when (val status = revocationData.revocationStatus) {
        is RevocationStatus.StatusList -> {
            "Index: ${status.idx}\nUrl: ${status.uri}"
        }
        is RevocationStatus.IdentifierList -> {
            "Identifier: ${status.id.toByteArray().toHex()}\nUrl: ${status.uri}"
        }
        else -> "Unknown revocation data format"
    }
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clickable {
                coroutineScope.launch {
                    val result = revocationChecker.check(
                        revocationStatus = revocationData.revocationStatus,
                        issuerCert = revocationData.certificate,
                        onlyTrusted = false
                    )
                    Logger.i(TAG, "RevocationCheckResult: $result")
                    val state = when (result.state) {
                        RevocationCheckState.VALID -> "Valid"
                        RevocationCheckState.INVALID -> "Invalid"
                        RevocationCheckState.SUSPENDED -> "Suspended"
                        RevocationCheckState.UNKNOWN -> "Unknown"
                    }
                    val trust = if (result.isTrusted) "Trusted" else "Not trusted"
                    statusText.value = if (result.error == null) {
                        "$state ($trust)"
                    } else {
                        "$state ($trust) [${result.error!!::class.simpleName}]"
                    }
                }
            }
    ) {
        FloatingItemHeadingAndText(
            heading = heading,
            text = "$detailsText\n${statusText.value}"
        )
    }
}

File: iosApp/iosApp/CredentialInfoScreen.swift
import SwiftUI
import Multipaz

struct CredentialInfoScreen: View {
    @Environment(ViewModel.self) private var viewModel
    let documentId: String
    let credentialId: String
    
    // State to show alerts for certificate details
    @State private var alertTitle = ""
    @State private var alertMessage = ""
    @State private var showAlert = false
    
    var body: some View {
        let documentInfo = viewModel.documentModel.documentInfos.first {
            $0.document.identifier == documentId
        }
        let credentialInfo = documentInfo?.credentialInfos.first {
            $0.credential.identifier == credentialId
        }
        
        ScrollView {
            VStack(alignment: .leading, spacing: 20) {
                Text("This screen contains low-level technical details about a specific credential")
                    .font(.footnote)
                if let credentialInfo = credentialInfo {
                    FloatingItemList {
                        FloatingItemHeadingAndText(
                            heading: "Type",
                            text: credentialInfo.credential.credentialType
                        )
                        FloatingItemHeadingAndText(
                            heading: "Identifier",
                            text: credentialInfo.credential.identifier
                        )
                        FloatingItemHeadingAndText(
                            heading: "Domain",
                            text: credentialInfo.credential.domain
                        )
                        FloatingItemHeadingAndText(
                            heading: "Certified",
                            text: credentialInfo.credential.isCertified ? "Yes" : "No"
                        )
                        
                        if credentialInfo.credential.isCertified {
                            FloatingItemHeadingAndDateTime(
                                heading: "Valid From",
                                dateTime: toSwiftDate(credentialInfo.credential.validFrom)
                            )
                            FloatingItemHeadingAndDateTime(
                                heading: "Valid Until",
                                dateTime: toSwiftDate(credentialInfo.credential.validUntil)
                            )
                            FloatingItemHeadingAndText(
                                heading: "Issuer provided data",
                                text: "\(credentialInfo.credential.issuerProvidedData.size) bytes"
                            )
                            FloatingItemHeadingAndText(
                                heading: "Usage Count",
                                text: credentialInfo.credential.usageCount.description
                            )
                            
                            // Revocation Status display
                            revocationSection(credential: credentialInfo.credential)
                            
                            // Specific credential types: MdocCredential
                            if let mdoc = credentialInfo.credential as? MdocCredential {
                                FloatingItemHeadingAndText(
                                    heading: "ISO mdoc DocType",
                                    text: mdoc.docType
                                )
                                FloatingItemHeadingAndText(
                                    heading: "ISO mdoc DS Key Certificate",
                                    text: "Click for details",
                                    trailingContent: {
                                        Image(systemName: "chevron.right").foregroundColor(.blue)
                                    }
                                )
                                .onTapGesture {
                                    showDsKeyCertDetails(credential: mdoc)
                                }
                            }
                            
                            // Specific credential types: SdJwtVcCredential
                            if let sdjwt = credentialInfo.credential as? SdJwtVcCredential {
                                FloatingItemHeadingAndText(
                                    heading: "Verifiable Credential Type",
                                    text: sdjwt.vct
                                )
                            }
                        }
                        
                        if let secureAreaBound = credentialInfo.credential as? SecureAreaBoundCredential {
                            FloatingItemHeadingAndText(
                                heading: "Secure Area",
                                text: secureAreaBound.secureArea.displayName
                            )
                            FloatingItemHeadingAndText(
                                heading: "Secure Area Identifier",
                                text: secureAreaBound.secureArea.identifier
                            )
                            if let keyInfo = credentialInfo.keyInfo {
                                FloatingItemHeadingAndText(
                                    heading: "Device Key Algorithm",
                                    text: keyInfo.algorithm.description_
                                )
                            }
                            FloatingItemHeadingAndText(
                                heading: "Device Key Invalidated",
                                text: credentialInfo.keyInvalidated ? "YES" : "No"
                            )
                            FloatingItemHeadingAndText(
                                heading: "Device Key Attestation",
                                text: "Click for details",
                                trailingContent: {
                                    Image(systemName: "chevron.right").foregroundColor(.blue)
                                }
                            )
                            .onTapGesture {
                                showDeviceKeyAttestationDetails(credential: secureAreaBound)
                            }
                        } else {
                            FloatingItemHeadingAndText(
                                heading: "Secure Area",
                                text: "N/A"
                            )
                        }
                    }
                    
                    if credentialInfo.credential.isCertified {
                        CredentialClaimsSection(credentialInfo: credentialInfo)
                    }
                } else {
                    FloatingItemList {
                        FloatingItemCenteredText(text: "No info for credential")
                    }
                }
            }
            .padding()
        }
        .navigationTitle("Credential info")
        .navigationBarTitleDisplayMode(.inline)
        .alert(isPresented: $showAlert) {
            Alert(
                title: Text(alertTitle),
                message: Text(alertMessage),
                dismissButton: .default(Text("OK"))
            )
        }
    }
    
    @ViewBuilder
    private func revocationSection(credential: Credential) -> some View {
        if let mdoc = credential as? MdocCredential,
           let revocationStatus = mdoc.mso.revocationStatus {
            if let statusList = revocationStatus as? RevocationStatus.StatusList {
                FloatingItemHeadingAndText(
                    heading: "Status List Revocation",
                    text: "Index: \(statusList.idx)\nUrl: \(statusList.uri)"
                )
            } else if let idList = revocationStatus as? RevocationStatus.IdentifierList {
                FloatingItemHeadingAndText(
                    heading: "Identifier List Revocation",
                    text: "Identifier: \(formatBytes(idList.id))\nUrl: \(idList.uri)"
                )
            } else if revocationStatus is RevocationStatus.Unknown {
                FloatingItemHeadingAndText(
                    heading: "Revocation info",
                    text: "Not parsed"
                )
            }
        }
    }
    
    private func toSwiftDate(_ instant: KotlinInstant) -> Date {
        return Date(timeIntervalSince1970: Double(instant.epochSeconds))
    }
    
    private func showDsKeyCertDetails(credential: MdocCredential) {
        Task {
            do {
                let dataSize = Int32(credential.issuerProvidedData.size)
                let issuerSigned = try Cbor.shared.decode(encodedCbor: credential.issuerProvidedData.toByteArray(startIndex: 0, endIndex: dataSize))
                let issuerAuth = issuerSigned.get(key: "issuerAuth").asCoseSign1
                
                let coseLabel = CoseNumberLabel(number: Cose.shared.COSE_LABEL_X5CHAIN)
                guard let certChainVal = issuerAuth.unprotectedHeaders[coseLabel] else {
                    showAlert(title: "Error", message: "Could not find x5chain in unprotected headers")
                    return
                }
                
                let certChain = certChainVal.asX509CertChain
                showAlert(title: "ISO mdoc DS Key Certificate", message: formatCertChain(certChain))
            } catch {
                showAlert(title: "Error", message: "Failed to extract certificate details: \(error.localizedDescription)")
            }
        }
    }
    
    private func showDeviceKeyAttestationDetails(credential: SecureAreaBoundCredential) {
        Task {
            do {
                let attestation = try await credential.getAttestation()
                if let certChain = attestation.certChain {
                    showAlert(title: "Device Key Attestation", message: formatCertChain(certChain))
                } else {
                    showAlert(title: "Device Key Attestation", message: "No attestation for Device Key")
                }
            } catch {
                showAlert(title: "Error", message: "Failed to get attestation: \(error.localizedDescription)")
            }
        }
    }
    
    private func formatCertChain(_ certChain: X509CertChain) -> String {
        let certs = certChain.certificates
        var result = "Certificate chain has \(certs.count) certificates:\n\n"
        for i in 0..<certs.count {
            let cert = certs[i]
            result += "Certificate [\(i)]:\n"
            result += "Subject: \(cert.subject)\n"
            result += "Issuer: \(cert.issuer)\n"
            result += "Validity: \(toSwiftDate(cert.validityNotBefore)) to \(toSwiftDate(cert.validityNotAfter))\n\n"
        }
        return result
    }
    
    private func showAlert(title: String, message: String) {
        self.alertTitle = title
        self.alertMessage = message
        self.showAlert = true
    }
    
    private func formatBytes(_ byteString: ByteString) -> String {
        let data = byteString.toNSData() as Data
        return data.map { String(format: "%02x", $0) }.joined()
    }
}

struct CredentialClaimsSection: View {
    let credentialInfo: CredentialInfo
    
    var body: some View {
        if let mdocClaims = credentialInfo.claims as? [MdocClaim] {
            let grouped = Dictionary(grouping: mdocClaims, by: { $0.namespaceName })
            let sortedNamespaces = grouped.keys.sorted()
            ForEach(sortedNamespaces, id: \.self) { namespace in
                let claims = grouped[namespace] ?? []
                FloatingItemList(title: "Namespace \(namespace)") {
                    ForEach(0..<claims.count, id: \.self) { index in
                        let claim = claims[index]
                        FloatingItemHeadingAndContent(
                            heading: claim.dataElementName,
                            content: { RenderClaimValue(claim: claim) }
                        )
                    }
                }
            }
        } else {
            let claims = credentialInfo.claims
            FloatingItemList(title: "Claims") {
                ForEach(0..<claims.count, id: \.self) { index in
                    let claim = claims[index]
                    let claimName = getClaimName(claim)
                    FloatingItemHeadingAndContent(
                        heading: claimName,
                        content: { RenderClaimValue(claim: claim) }
                    )
                }
            }
        }
    }
    
    private func getClaimName(_ claim: Claim) -> String {
        if let jsonClaim = claim as? JsonClaim {
            return jsonClaim.claimPath.map { $0.jsonPrimitive.content }.joined(separator: ".")
        } else {
            return claim.displayName
        }
    }
}
